### PR TITLE
Fix #2619 - When redis feed is empty, fall back to database

### LINKED ISSUE #2619: Empty home timeline after period of inactivity
Didn't log in for a week. Came back and my Home timeline was empty. Also seen on Tusky.

Whatever mechanism caches a person's Followed toots doesn't seem to work if not actively logging in.

* * * *

- [X] I searched or browsed the repo’s other issues to ensure this is not a duplicate.
- [X] This bug happens on a [tagged release](https://github.com/tootsuite/mastodon/releases) and not on `master` (If you're a user, don't worry about this).

### PR DESCRIPTION
**Pull request by Gargron**
Home feeds are normally stored in redis. But removed for users inactive for > 14 days.

When user logs back in, a background worker starts to regenerate the feed. However, before that job completes, the user is greeted with a blank home timeline.

This change will fallback to database queries if RegenerationWorker has been queued up but hasn't finished running.

### CODE REVIEW

File: app/models/feed.rb [hunk 1]
```diff
@@ -7,15 +7,28 @@ def initialize(type, account)
...
     since_id   = '-inf' if since_id.blank?
     unhydrated = redis.zrevrangebyscore(key, "(#{max_id}", "(#{since_id}", limit: [0, limit], with_scores: true).map(&:last).map(&:to_i)
     status_map = Status.where(id: unhydrated).cache_ids.map { |s| [s.id, s] }.to_h
-
     unhydrated.map { |id| status_map[id] }.compact
   end
 
-  private
+  def from_database(limit, max_id, since_id)
+    results = Status.as_home_timeline(@account).paginate_by_max_id(limit, max_id, since_id).to_a
```
Comment (akihikodaki) [inline app/models/feed.rb]: You can use `reject` instead of `to_a` and `reject!`.

**Review by akihikodaki — CHANGES_REQUESTED** [review]
I request a small minor change; please see the review.
By the way, I do not think the code to regenerate feeds are spread out across modules (`UserTrackingConcern`, `Feed`, and `PrecomputeFeedService`), but that can be resolved later.

**Comment by clworld** [comment]
Isn't it good also fallback to database when user reached end of feed(400)?
In public/local timeline user can read history infinitely, but not in home timeline. User can read only 400 recent toots.

File: app/models/feed.rb (hunk 1 above)
Comment (clworld) [inline app/models/feed.rb]: It seems Web UI will stop paging when reject! below filters all of statuses in page.
But it's OK because it's used only when Feed is not created.

Comment (Gargron) [inline app/models/feed.rb]: @akihikodaki I wanted to do it this way to avoid copying the array (reject! mutates existing array afaik)

@clworld Yes, that's a problem. Maybe limit should be doubled? Or do second request if the result of reject! is empty.. But even then there is no guarantee new items won't be rejected.

I am reminded that @akihikodaki wrote a version of the home query that includes the rules from FeedManager#filter?. Perhaps this is the place to dig that up and use it...

Comment (akihikodaki) [inline app/models/feed.rb]: > I wanted to do it this way to avoid copying the array (reject! mutates existing array afaik)

`to_a` copies an internal array, so the number of copy is same. The below is the long explanation.

Here is the implementation of `to_a`:
```Ruby
# Converts relation objects to Array.
def to_a
    records.dup
end
```
https://github.com/rails/rails/blob/v5.1.1/activerecord/lib/active_record/relation.rb#L249

`reject` is defined in `Enumerable`. `reject` itself copies the source, but there is no additional copy. The source is the very array copied in `to_a`.
> The class must provide a method `each`
https://docs.ruby-lang.org/en/2.4.0/Enumerable.html

```Ruby
    delegate :to_xml, :encode_with, :length, :each, :uniq, :to_ary, :join,
         :[], :&, :|, :+, :-, :sample, :reverse, :compact, :in_groups, :in_groups_of,
         :to_sentence, :to_formatted_s, :as_json,
         :shuffle, :split, :index, to: :records
```
https://github.com/rails/rails/blob/v5.1.1/activerecord/lib/active_record/relation/delegation.rb

In summary, both of `to_a` and `reject` are copying the same array once.

Comment (Gargron) [inline app/models/feed.rb]: @akihikodaki Done!

**Review by clworld — APPROVED** [review]
Looks good to me.

File: app/controllers/concerns/user_tracking_concern.rb [hunk 2]
```diff
@@ -27,4 +27,9 @@ def user_needs_sign_in_update?
   def user_needs_feed_update?
     current_user.last_sign_in_at < REGENERATE_FEED_DAYS.days.ago
   end
+
+  def regenerate_feed!
+    Redis.current.setnx("account:#{current_user.account_id}:regeneration", true) == 1 && Redis.current.expire("account:#{current_user.account_id}:regeneration", 3_600 * 24)
```
Comment (clworld) [inline app/controllers/concerns/user_tracking_concern.rb]: SETNX + EXPIRE can be performed by one redis command. https://redis.io/commands/set
Redis-rb seems support them. https://github.com/redis/redis-rb/blob/master/lib/redis.rb#L735

Comment (clworld) [inline app/controllers/concerns/user_tracking_concern.rb]: Redis.current.set("account:#{current_user.account_id}:regeneration", true, nx: true, ex: 3_600 * 24)

**Review by akihikodaki — APPROVED** [review]
